torch/csrc/jit/autodiff.cpp

```diff
@@ -319,107 +273,105 @@ static void lambdaLiftReverse(Graph& graph,
     // incorrect sum that doesn't use prev vjp, replace uses, and fix the sum.
-    Value * new_vjp = addValues(tmp_vjp_in, tmp_vjp_in, tmp_vjp_prev->node());
+    Value * new_vjp = toVar(tmp_vjp_in) + toVar(tmp_vjp_in);
+    new_vjp->node()->moveAfter(tmp_vjp_prev->node());
     tmp_vjp_prev->replaceAllUsesWith(new_vjp);
     new_vjp->node()->replaceInput(1, tmp_vjp_prev);
     grad_desc.df_input_vjps.emplace_back(i);
   }
 
-  // -- Construct reverse_inputs -----------------------------------------------
-  // Quick reference:
-  //   [captured primal values, in topological order],            1st loop below
-  //   [output vjps (aka grad_outputs)], [temporary vjps]         2nd loop below
-  value_list reverse_inputs;
-  for (Capture capture : grad_desc.df_input_captures) {
-    auto & source = capture.kind == Capture::Kind::Input ? primal_inputs : primal_outputs;
-    reverse_inputs.push_back(source[capture.offset]);
-  }
-  // These are the vjps computed by differentiate + the code above
-  for (Value * reverse_vjp : filter(graph.inputs(), is_stage_1))
-    reverse_inputs.push_back(reverse_vjp);
-
-  // Finally, we can split the graph into two parts.
-  grad_desc.f  = splitOffStage(graph, 0, primal_inputs, primal_outputs);
-  grad_desc.df = splitOffStage(graph, 1, reverse_inputs, reverse_outputs);
+  // add the captures as formal arguments to the reverse_block
+  // afterward inputs: [output vjps][temporary vjps][captures]
+  // construct a map from captured 'value' to the index in the input list
+  // used to extract this block into its own function
+  std::unordered_map<Value*, size_t> capture_to_formal_index;
+  const auto & add_capture = [&](Value * captured) {
+    capture_to_formal_index[captured] = reverse_block->inputs().size();
+    reverse_block->addInput()->copyMetadata(captured);
+  };
+  for(auto & offset : grad_desc.df_input_captured_inputs)
+    add_capture(graph.inputs()[offset]);
+  for(auto & offset : grad_desc.df_input_captured_outputs)
+    add_capture(graph.outputs()[offset]);
+
+  grad_desc.df = std::make_shared<Graph>();
+  grad_desc.df->block()->cloneFrom(reverse_block, [&](Value* v) {
+    return grad_desc.df->inputs()[capture_to_formal_index.at(v)];
+  });
+  // reverse_node was just to hold onto reverse_block in a debuggable way
+  // we can remove it now.
+  reverse_block->owningNode()->destroy();
 }
 
 Gradient differentiate(std::shared_ptr<Graph>& _graph, const std::vector<bool>& requires_grad) {
+  Gradient grad_desc;
   // Take ownership of the graph
-  std::shared_ptr<Graph> graph;
   JIT_ASSERTM(_graph.use_count() == 1,
               "differentiate will mutate and destroy the graph, so it requires "
               "graph.use_count() == 1");
-  std::swap(_graph, graph);
+  std::swap(_graph, grad_desc.f);
   // XXX: Take care when handling outputs - they can be duplicated!
-  Gradient grad_desc;
+
+  WithInsertPoint guard(*grad_desc.f, grad_desc.f->block());
```

Comment:
Why is this insert point necessary? Isn't the main block a default insert point?